command/agent/cache/api_proxy.go

```diff
@@ -0,0 +1,61 @@
+package cache
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/api"
+)
+
+// APIProxy is an implementation of the proxier interface that is used to
+// forward the request to Vault and get the response.
+type APIProxy struct {
+	logger hclog.Logger
+}
+
+type APIProxyConfig struct {
+	Logger hclog.Logger
+}
+
+func NewAPIProxy(config *APIProxyConfig) Proxier {
+	return &APIProxy{
+		logger: config.Logger,
+	}
+}
+
+func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, error) {
+	client, err := api.NewClient(api.DefaultConfig())
```

Comment:
I don't think you should create a new API client for every single call. This then requires creating a new HTTP transport/pool and so on. I think it should be sufficient to use Clone().

Reply:
This was discussed between Chris and I and we had concluded on just using a new client. That thought was primarily based on the assumption that creating a new client is not going to be heavy. But, I'll work on Clone and see how that pans out.

Reply:
If clients are using pooled connections you could pretty easily exhaust open file sockets if you are creating new connection pools for every connection. I think. So I do think this could be heavier than you think.

Reply:
Lots of thinking in that comment. Thoughts.

Reply:
Agreed. I hadn't thought of exhausting file sockets. This needs to be changed. I'm working on this.

Reply:
Being tracked in https://github.com/hashicorp/vault/pull/6257/files